src/chromatix/functional/propagation.py

```diff
@@ -274,32 +253,123 @@ def asm_propagate(
             for zero padding.
         kykx: If provided, defines the orientation of the propagation. Should
             be an array of shape `[2,]` in the format `[ky, kx]`.
+        remove_evanescent: If ``True``, removes evanescent waves. Defaults to
+            False.
         bandlimit: If ``True``, bandlimited the kernel according to "Band-
             Limited Angular Spectrum Method for Numerical Simulation of Free-
             Space Propagation in Far and Near Fields" (2009) by Matsushima and
             Shimobaba. Defaults to ``False``.
         shift_yx: If provided, defines a shift in microns in the destination
             plane. Should be an array of shape `[2,]` in the format `[y, x]`.
+        dx: If provided, defines a different output sampling at the output
+            plane.
+        N_out: If provided, defines the output shape of the field. Should be
+            a tuple of integers. If not provided and ``dx`` is provided, the
+            output shape will default to that of the input field.
+        use_czt: Whether or not to use chirp Z-transform for off-axis
+            propagation or interpolating with different output sampling.
+            Defaults to True if ``dx`` or ``N_out`` is provided.
         mode: Either "full" or "same". If "same", the shape of the output
             ``Field`` will match the shape of the incoming ``Field``. Defaults
             to "full", in which case the output shape will include padding.
     """
     field = pad(field, N_pad, cval=cval)
-    propagator = compute_asm_propagator(field, z, n, kykx, bandlimit, shift_yx)
-    field = kernel_propagate(field, propagator)
+    if dx is None and N_out is None:
+        # If neither dx nor N_out is provided, use the default ASM propagation
+        # as FFT is faster than CZT
+        use_czt = False
+    propagator = compute_asm_propagator(
+        field,
+        z,
+        n,
+        kykx,
+        bandlimit,
+        shift_yx if not use_czt else (0.0, 0.0),
+        remove_evanescent=remove_evanescent,
+    )
+    field = kernel_propagate(
+        field, propagator, dx=dx, N_out=N_out, shift_yx=shift_yx, use_czt=use_czt
+    )
     if mode == "same":
         field = crop(field, N_pad)
     return field
 
 
-def kernel_propagate(field: Field, propagator: Array) -> Field:
+def kernel_propagate(
+    field: Field,
+    propagator: Array,
+    dx: Union[float, Array] = None,
+    N_out: Tuple[int, int] = None,
+    shift_yx: Union[Array, Tuple[float, float]] = (0.0, 0.0),
+    use_czt: bool = False,
+) -> Field:
     """
     Propagate an incoming ``Field`` by the given propagation kernel
     (``propagator``). This amounts to performing a Fourier convolution of the
     ``field`` and the ``propagator``.
     """
     axes = field.spatial_dims
-    u = ifft(fft(field.u, axes=axes) * propagator, axes=axes)
+    if dx is None and N_out is None and not use_czt:
+        u = ifft(fft(field.u, axes=axes) * propagator, axes=axes)
```

Comment:
Looks like shifting is not handled in this case (this is basically what we had before this rescaling update). I guess we should be able to do https://opg.optica.org/oe/fulltext.cfm?uri=oe-18-17-18453&id=205150 here?

Reply:
So yeah this might deserve a comment because the shift in this case is taken into account when computing `propagator` (as is done in that paper): https://github.com/ebezzam/chromatix/blob/c65eeda603161f590f6129c7bb2962d382423493/src/chromatix/functional/propagation.py#L287

Reply:
Ah, makes sense!